pkg/kptfile/pkgfile.go

```diff
@@ -58,6 +58,39 @@ type KptFile struct {
 
 	// Parameters for inventory object.
 	Inventory *Inventory `yaml:"inventory,omitempty"`
+
+	// rNode is the yaml.RNode representation of Kptfile
+	rNode *yaml.RNode `yaml:"inventory,omitempty"`
```

Comment:
I'm not sure I agree with the direction of this change. It's conflating the idiomatic Go structures of the Kptfile schema with low-level YAML marshalling and unmarshalling. Instead you want the following:

1) Use the Go structure throughout the code base. BTW, the new schema is being defined here: https://github.com/GoogleContainerTools/kpt/pull/1417/files
When a feature in kpt needs to work with Kptfile, they should just be able to use the struct instead of traversing a YAML AST every time.
2) Marshalling and unmarshing to YAML while preserving comments (and perhaps ordering) should be done as the first and last mile operation and encapsulated by a read and write io utilities. 

Another way to think about this: We should be able to formulate logic in step 1 as a kpt function if we wanted to do. Like any other KRM resource, functions should be able to manipulate the Kptfile without using kyaml or traversing the YAML ast. Today, we can use develop functions in idiomatic javascript or python (round trip through their native structures) to operate on KRM resources. The functions don't need to worry about YAML at all or preserving comments, etc. kpt orchestrator (step 2) takes care of all that complexity instead of pushing the burden to each function implementation.

Reply:
There are 2 problems with that approach.
1. What if a user wants to add additional package metadata to Kptfile which is not known to kpt. unmarshalling and marshalling into types would loose that information.
2. If users change the order of the fields, we would be rewriting that order which leads to unwanted diffs.

Reading and writing Kptfile as any other KRM resource, would solve those problems. It also makes it simple to merge Kptfiles similar to other KRM resources.

Developers will still be using the Go structure everywhere in the code. But instead of setting them directly on the Kptfile struct, we would offer encapsulation e.g. kf.SetUpstream(upstream Upstream) so that the low level library would handle the yaml AST traversal.

Reply:
1What's the use case driving 1? Kptfile has both `metadata.[label,annotation]` and `packageMeta.tags` for humans and automation to add additoinal metadata?

2 and 3 are general problems for any mutator functions, not specific to kpt CLI mutating a particular type called Kptfile. 
Users should also be able to develop meta functions that mutate Kptfile, so we need to handle it in a general way.
Preservation of comments is already handle by the orchestrator: We have functions written in TS that lose comments by roundtripping through Javascript objects. Kpt is smart enough to preserve comments at a higher level when writing back to disk. Same should apply to ordering of fields. Although I talked to Mortent about this: we may want to think about having ordering dictated by 'kpt cfg fmt' if we consider them a styling issue (Similar to how `go fmt` removed all the flamewar about style issues by just dictating the format).

"Reading and writing Kptfile as any other KRM resource, would solve those problems. It also makes it simple to merge Kptfiles similar to other KRM resources."

This is taking a very kyaml-specific view of the world. Instead, we need to generalize this even further. Kptfile is not special. Logic that manipulates Kptfile is not special (Don't assume it should only be done using kyaml). We need to be able to write functions in different toolchains and still preserve comments, provide consistent ordering (whether preserving or dictating), etc.

"Developers will still be using the Go structure everywhere in the code. But instead of setting them directly on the Kptfile struct, we would offer encapsulation e.g. kf.SetUpstream(upstream Upstream) so that the low level library would handle the yaml AST traversal."

Can you instead formulate this as such:

1. read -> Kptfile (Go struct)
2. Mutate Kptfile struct as normal
3. write(Kptfile)

All the complexity is encapsulated in step 3. You can traverse the yaml AST and compare it to fields in the Kptfile struct, and manipulate the ast so you can preserve commetns, etc.

Reply:
So I think there are some good points for both approaches, and I think this is bigger than just how we manage the Kptfile. We should think about what our goals are here. I think there are a few fundamental principles we want to follow:
- To the extent possible and practical, we want to treat the Kptfile the same as other KRM resources.
- We should be able to support multiple languages and tools
- The source of truth for the API of the Kptfile (and other KRM resources) is the openapi schema, not the implementation in any particular language.
- Fields within KRM resources should not be reordered in a way that is visible to users. Dictating a specific ordering might be acceptable for the Kptfile, it is not ok for KRM resources in general.

Some consequences of this as it relates to the comments here are:

1. We should be careful about unmarshalling the Kptfile into structs/objects in languages like Go and Java, in particular in functions. If we ever want to add fields to the Kptfile (which I think we want to be able to do without having to bump the api version), we need functions written against older versions of the API to work without dropping fields. This is probably less of an issue with Javascript and Python if the Kptfile is unmarshalled into objects or dicts.

2. If we want to handle ordering as a last mile operation, we need to verify that we can do this well. Doing this when functions can both reorder fields and makes changes, this doesn't seem trivial. I'm a little worried that the resulting output might differ depending on which language/tool is used in the functions.

3. Handling comments purely as a first/last mile operation doesn't seem feasible for all use-cases. Setter refs are comments, so any functions that uses setters must be able to handle line comments. While the setter function will probably we written in Go, the resources passed into it must have the comments intact.

So it seems like the orchestrator will have to "fix" ordering and comments between each function in the pipeline rather than just at the end. Looking at the function code in the kustomize repo, it seems like we are already doing this today.

As for whether we let the Kptfile be backed by a kyaml structure and handle ordering/comments by avoiding unmarshalling to structs, or we unmarshall to go structs and clean it up during write, I don't really feel strongly either way as long as it follows the principles above. I thought using a kyaml structure would be easier, but I might be wrong.

The biggest risk I see here, is that we don't yet know how well we can fix ordering in the orchestrator. We should try this out since it is important for supporting non-golang functions.

Reply:
I mostly agree, but would expand on a few things:

1. Re "We should be able to support multiple languages and tools"

Enable functions to be written idiomatically in the respective languages. If developing in Javascript, developer should use the native Javascript object. In Golang, developer should be able to use Go structs to take advantage of the Go compiler to provide type safety. Using low-level YAML AST should only be needed if the function is interested in YAML-only concepts like comments.

2. "The source of truth for the API of the Kptfile (and other KRM resources) is the openapi schema, not the implementation in any particular language."

Yes. However, the following are isomorphic: Go struct, CRD, OpenAPI schema. So we can author one and auto-generate the other two. This is purely an implementation choice. Also keep in mind that we're going need both CRD and the OpenAPI schema since other tooling may require the CRD. 

What's not an implementation detail: Kptfile needs to be treated like any other custom resource by the KRM tooling ecosystem. This means, Kptfile should follow the same versioning semantics as any other custom resource. For example, Cloud Code should be able to render and validate the Kptfile similar to how it treats a Deployment or a custom resource like ResourceHierarchy. If we're changing the schema, we need to think about the ecosystem of tooling, and not just how the Kpt CLI will be impacted.

3. Implementation detail of Kpt CLI. Type safety is good and eliminates entire class of bugs, making development faster and cheaper. To have static typing, we need Go structs. Otherwise, a huge reason for choosing a language like Go for development becomes moot. This also follows from 1 if you just think of most of the functionality in Kpt CLI like any other kpt function (This is where separating the orchestrator from function comes into play, or).

Reply:
Hopefully without derailing this vibrant discussion I have a couple of questions.

IMHO it's definitely worthwhile to get the "data" from the Kptfile which helps us execute the right instructions.  However it also seems we take those instructions and in several places end up writing those instructions back to disk potentially overwriting what we originally had.  I am just looking at all the instances of `WriteFile` being called: `hardResetSourceFiles`, `UpsertKptfile` etc.   

Do we want to do a wholesale writeout of the Kptfile struct or is it better to just do a surgical update of the file on disk?
Is it better to not conflate these two things?  The execution instructions can live in a Kptfile struct which is strongly typed and defined - after all we can only execute what we know.   The update to the Kptfile on disk would be handled in a different way?